### DISCUSSION TITLE
[Argo] UI plugins

### DISCUSSION THREAD

**Discussion by nate-double-u** [cncf/mentoring#567] [Google Summer of Code 2022]
Currently, any user who wants to add additional resources to the UI needs to implement both backend and frontend changes. Instead, we'd like to implement an UI extension mechanism to load and embed UI elements in Argo Workflows UI. Argo CD already has a UI extension mechanism to load the Argo Rollouts into the UI that is proven to be successful. We can lift inspiration from this and implement something similar in Argo Workflows.

Mentor(s): Alex Collins (@alexec)

For more information, please see: https://github.com/cncf/mentoring/blob/main/summerofcode/2022.md#ui-plugins

**Comment by shubham4443** [comment]
Hi @alexec!
I'm interested in applying at GSoC for this project. I have started early contribution in argo-workflows ([argo/argo-workflows#8208](https://github.com/argoproj/argo-workflows/pull/8208)) to get to know and understand the structure and flow of codebase better. I have couple of questions.

1. You have mentioned to take inspiration from argocd. Arogcd implements a CRD for UI Extension. Should we do the same for argo-workflows?
2. Can you give a couple of use cases as an user/operator where he/she would have to make use of UI Extension in argo workflows?
3. In addition to [this](https://google.github.io/gsocguides/student/writing-a-proposal), do you expect anything more in proposal?

Thanks!

**Comment by alexec** [comment]
Hi @shubham4443 ,

Great to hear your are interested!

1. Argo CD CRD

Having spent some time working with the Argo CD extension mechanism, I’m not sure we want or need a CRD. It’s would be large amount of work to implement, but does not work in air-gapped systems. In Argo CD, extensions are installed in /tmp/extensions. All we would need to way to get them there, and that could easily be done with a volume mount. Works in air-gapped systems. No CRD needed.

2. Use Cases

The two kinds of extension I hope for are:

* An extension that add CI capabilities to the UI. For example, the ability to view test reports (e.g. JUnit XML reports) in the UI. 
* ML extension. This is more of a blank canvas. One idea would be the ability to author a workflow in the UI using either Hera or Couler. That would be an editor, or could be a Jupiter notebook. 

3. Addition to this…

No. I think that looks fine to me.

**Comment by shubham4443** [comment]
Hi @alexec!

I have prepared an initial draft of the proposal and would like to get your reviews on it. Should I mail you the proposal or we can continue the conversation on slack?

**Reply by alexec** [reply]
I suggest email for something important like this

**Reply by shubham4443** [reply]
Done. Let me know if you didn't receive the mail.
